File: game/scripts/vscripts/dota2/net_table.lua
local M = require("pl.class")()

function M:_init(name)
  self.name = name
end

function M:Set(key, value)
  CustomNetTables:SetTableValue(self.name, key, value)
end

function M:Get(key, defaultValue)
  local value = CustomNetTables:GetTableValue(self.name, key)

  if value == nil then
    return defaultValue
  end

  return value
end

return M

File: content/panorama/scripts/custom_game/lib/net_table.js
"use strict";

((global /*, context */) => {
  const { lodash: _ } = global;

  class NetTable {
    constructor(name) {
      this.name = name;

      this._onChangeCallbacks = [];
      this._onKeyChangeCallbacks = {};

      this.subscribe(this.onChange.bind(this));
    }

    onChange(table, key, value) {
      if (table !== this.name) {
        return;
      }

      _.over(this._onChangeCallbacks)(key, value);
      _.over(_.get(this._onKeyChangeCallbacks, key, []))(key, value);
    }

    subscribe(fn) {
      return CustomNetTables.SubscribeNetTableListener(this.name, fn);
    }

    All() {
      return CustomNetTables.GetAllTableValues(this.name);
    }

    Get(key) {
      return CustomNetTables.GetTableValue(this.name, key);
    }

    OnChange(fn) {
      this._onChangeCallbacks.push(fn);
    }

    OnKeyChange(key, fn) {
      this._onKeyChangeCallbacks[key] = this._onKeyChangeCallbacks[key] || [];
      this._onKeyChangeCallbacks[key].push(fn);
    }
  }

  global.NetTable = NetTable;
})(GameUI.CustomUIConfig(), this);
